cirq/sim/mps_simulator.py

```diff
@@ -0,0 +1,350 @@
+                )
+            )
+
+        return np.array(measurements, dtype=int)
+
+
+@value.value_equality
+class MPSState:
+    """A state of the MPS simulation."""
+
+    def __init__(self, qubit_map, initial_state=0):
+        self.qubit_map = qubit_map
+        self.M = []
+        for qubit in qubit_map.keys():
+            d = qubit.dimension
+            x = np.zeros(
+                (
+                    1,
+                    1,
+                    d,
+                )
+            )
+            x[0, 0, (initial_state % d)] = 1.0
+            self.M.append(x)
+            initial_state = initial_state // d
+        self.M = self.M[::-1]
+        self.threshold = 1e-3
+
+    def __str__(self) -> str:
+        return str(self.M)
+
+    def _value_equality_values_(self) -> Any:
+        return self.qubit_map, self.M, self.threshold
+
+    def copy(self) -> 'MPSState':
+        state = MPSState(self.qubit_map)
+        state.M = [x.copy() for x in self.M]
+        state.threshold = self.threshold
+        return state
+
+    def state_vector(self):
+        M = np.ones((1, 1))
+        for i in range(len(self.M)):
+            M = np.einsum('ni,npj->pij', M, self.M[i])
+            M = M.reshape(M.shape[0], -1)
+        assert M.shape[0] == 1
+        return M[0, :]
+
+    def to_numpy(self) -> np.ndarray:
+        return self.state_vector()
+
+    def apply_unitary(self, op: 'cirq.Operation'):
+        idx = [self.qubit_map[qubit] for qubit in op.qubits]
+        U = protocols.unitary(op).reshape([qubit.dimension for qubit in op.qubits] * 2)
+
+        if len(idx) == 1:
+            n = idx[0]
+            self.M[n] = np.einsum('ij,mnj->mni', U, self.M[n])
+        elif len(idx) == 2:
+            n = idx[0]
+            p = idx[1]
+            if abs(n - p) != 1:
+                raise ValueError('Can only handle continguous qubits')
+            T = np.einsum('klij,mni,npj->mkpl', U, self.M[n], self.M[p])
+            X, S, Y = np.linalg.svd(T.reshape([T.shape[0] * T.shape[1], T.shape[2] * T.shape[3]]))
+            X = X.reshape([T.shape[0], T.shape[1], -1])
+            Y = Y.reshape([-1, T.shape[2], T.shape[3]])
+
+            S = np.asarray([math.sqrt(x) for x in S])
+
+            nkeep = 0
+            for i in range(S.shape[0]):
+                if S[i] >= S[0] * self.threshold:
+                    nkeep = i + 1
+
+            X = X[:, :, :nkeep]
+            S = np.diag(S[:nkeep])
+            Y = Y[:nkeep, :, :]
+
+            self.M[n] = np.einsum('mis,sn->mni', X, S)
+            self.M[p] = np.einsum('ns,spj->npj', S, Y)
+        else:
+            raise ValueError('Can only handle 1 and 2 qubit operations')
+
+    def perform_measurement(
+        self, qubits: Sequence[ops.Qid], prng: np.random.RandomState, collapse_state_vector=True
+    ) -> List[int]:
+        results: List[int] = []
+
+        if collapse_state_vector:
+            state = self
+        else:
+            state = self.copy()
```

Comment:
To confirm: this copy is required for non-collapsing measurement because `state` must be mutated to capture how each qubit measurement affects the rest (e.g. measuring `0` on the first qubit of state `|00) + |11)` forces a `0` on the second). Is this correct?

Reply:
I think so. There's test_simulate_moment_steps_sample that makes sure that this is the intended behavior and I was inspiring myself from this:
https://sourcegraph.com/github.com/quantumlib/Cirq@9bad962c32a236ba72d829e5091d3bbc7711814c/-/blob/cirq/sim/clifford/clifford_simulator.py#L360

Leaving unresolved.

Reply:
Sounds good to me. If I understand it correctly, the MPS simulator has much smaller "state" objects (similar to the CliffordSimulator) so this is less of a concern.

Reply:
I think the size of the state here depends on the degree of entanglement, but if it's too big, then it probably doesn't make sense to use the proposed simulator in the first place.